### PR TITLE
List view comments

### PR DESCRIPTION
**Pull request by mdbenjam**
Update the table on the list view page to see comments. Three main changes here:
1. Change the Table component to allow different rows to have different numbers of columns
1. Add expandable rows of comments to the document list view
1. Update how we scroll to comments

connects: #1268 #1450

**Test Plan**
- [x] Verify that all comments are viewable on the list view page
- [x] Verify that when 'jump to section' is clicked, it jumps into the document and scrolls to the comment
- [x] Run new tests

### CODE REVIEW

File: app/assets/stylesheets/_main.scss [hunk 1]
```diff
@@ -241,6 +241,17 @@ select {
   }
 }
 
+.cf-no-padding-btn-link {
```
Comment (mdbenjam) [inline app/assets/stylesheets/_main.scss]: I don't like this at all... The current cf-btn-link has !importants on padding-top and padding-bottom. And the styling is all in caseflow commons. I'm trying to remove the important. That will come in another PR.

File: client/app/util/AnnotationStorage.js [hunk 2]
```diff
@@ -34,20 +34,19 @@ export default class AnnotationStorage {
...
       let data = ApiUtil.convertToSnakeCase({ annotation });
 
       ApiUtil.post(`/document/${documentId}/annotation`, { data }).
           then((response) => {
-
+            let allAnnotations = this.getAnnotationByDocumentId(documentId);
             let responseObject = JSON.parse(response.text);
 
             annotation.uuid = responseObject.id;
+            annotation.id = responseObject.id;
```
Comment (mdbenjam) [inline client/app/util/AnnotationStorage.js]: We need to set the returned id before adding it to the list of annotations.

File: client/app/components/Table.jsx [hunk 3]
```diff
@@ -60,12 +60,28 @@ export default class Table extends React.Component {
...
 
+    let getCellSpan = (rowObject, column) => {
+      if (column.span) {
+        return column.span(rowObject);
+      } else {
+        return 1;
+      }
+    }
+
+    let getColumns = (props) => {
```
Comment (mdbenjam) [inline client/app/components/Table.jsx]: @shanear What do you think of this modification to tables? I need a way to show row dependent content. I could also make a new component if this is getting too complicated.

Comment (shanear) [inline client/app/components/Table.jsx]: Is this because there are some rows that need multiple colspans?

Comment (mdbenjam) [inline client/app/components/Table.jsx]: Exactly

File: client/app/components/Comment.jsx [hunk 4]
```diff
@@ -16,33 +16,77 @@ export default class Comment extends React.Component {
...
-            </span>
-            <Button
-              name="delete"
-              classNames={["cf-btn-link comment-control-button"]}
-              onClick={this.onDeleteComment}>
-              Delete
-            </Button>
+            {this.controlButtons()}
           </span>
+          <b>Page {this.props.page} {jumpToSectionButton}</b>
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: How about the more semantic `<strong>` instead of `<b>`?

File: client/app/components/Comment.jsx [hunk 5]
```diff
@@ -16,33 +16,77 @@ export default class Comment extends React.Component {
...
+        </div>
+        <div
+          className={className}
+          key={this.props.children.toString()}
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: Why is this needed?

File: client/app/components/Comment.jsx [hunk 6]
```diff
@@ -51,15 +95,25 @@ export default class Comment extends React.Component {
...
         </div>
       </div>;
+    }
+
+    return commentToRender;
   }
 }
 
+Comment.defaultProps = {
+  onClick: () => { /* do nothing */ }
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: How about:

```js
onClick: _.noop
```

File: client/app/components/Table.jsx [hunk 7]
```diff
@@ -37,10 +37,18 @@ export default class Table extends React.Component {
       return alignmentClasses[column.align];
     };
 
+    let getColumns = (props) => {
+      if (typeof props.columns === 'function') {
```
Comment (NickHeiner) [inline client/app/components/Table.jsx]: This can also be:

```js
if (_.isFunction(props.columns)) {
```

Comment (NickHeiner) [inline client/app/components/Table.jsx]: I would probably do this function as a one-line ternary statement but it's up to you.

File: client/app/reader/DecisionReviewer.jsx [hunk 8]
```diff
@@ -84,7 +85,19 @@ export default class DecisionReviewer extends React.Component {
     });
   }
 
-  showPdf = (pdfNumber) => (event) => {
+  pdfNumberFromId = (pdfId) => {
```
Comment (NickHeiner) [inline client/app/reader/DecisionReviewer.jsx]: How about:

```js
pdfNumberFromId = (pdf) => _.findIndex(this.state.documents, { id: pdfId })
```

File: client/app/reader/PdfListView.jsx [hunk 9]
```diff
@@ -33,85 +56,122 @@ export default class PdfListView extends React.Component {
...
-          </span>;
-        }
+    return (row) => {
+      if (row && row.isComment) {
+        return [{
+          valueFunction: (doc) => {
+            let comments = this.props.annotationStorage.
+              getAnnotationByDocumentId(doc.id);
+            let commentNodes = comments.map((comment, commentIndex) => {
+              return <Comment
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: Doesn't this element need a `key`? 

If this is a list of content, it could be good to wrap it in a `ul` for accessibility.

File: client/app/reader/PdfListView.jsx [hunk 10]
```diff
@@ -33,85 +56,122 @@ export default class PdfListView extends React.Component {
...
+                horizontalLayout={true}>
+                  {comment.comment}
+                </Comment>;
+            });
+
+            return <div>
+              {commentNodes}
+            </div>;
+          },
+          span: () => {
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: This can also be:

```js
span: _.constant(NUMBER_OF_COLUMNS)
```

File: client/app/reader/PdfListView.jsx [hunk 11]
```diff
@@ -33,85 +56,122 @@ export default class PdfListView extends React.Component {
...
+        }];
       }
-    ];
+
+      return [
+        {
+          header: <div
+            id="categories-header"
+            className="document-list-header-categories"
+            onClick={() => {
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: I would just remove this until we actually have a click handler.

Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: (I know you didn't add this in this PR.)

File: client/app/reader/PdfListView.jsx [hunk 12]
```diff
@@ -123,6 +183,18 @@ export default class PdfListView extends React.Component {
       commentSelectorClassNames.push('cf-label');
     }
 
+    let rowObjects = this.props.documents.reduce((acc, row) => {
+      acc.push(row);
+      if (this.state.commentsOpened[row.id]) {
+        let commentRow = { ...row };
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: How about:

```js
const commentRow = { ...row, isComment: true};
```

**Comment by NickHeiner** [comment]
I've added a few minor comments. Overall it looks good. I will continue to look further when I get a chance.

File: client/app/components/Table.jsx (hunk 3 above)
Comment (mdbenjam) [inline client/app/components/Table.jsx]: Yes indeed

File: client/app/components/Comment.jsx [hunk 13]
```diff
@@ -98,8 +102,10 @@ export default class Comment extends React.Component {
 }
 
 Comment.defaultProps = {
-  onClick: () => {}
-}
+  onClick: () => {
+    return _.noop;
```
Comment (mdbenjam) [inline client/app/components/Comment.jsx]: If we don't return, the linter complains about this being an unused expression...

Comment (NickHeiner) [inline client/app/components/Comment.jsx]: That makes sense. This should be:

```js
onClick: _.noop
```

File: app/assets/stylesheets/_main.scss [hunk 14]
```diff
@@ -241,6 +241,17 @@ select {
...
+  background: transparent;
+  color: $color-primary;
+  display: inline-block;
+  font-weight: normal;
+  line-height: 1;
+  -webkit-font-smoothing: inherit;
+  padding-top: 0;
```
Comment (NickHeiner) [inline app/assets/stylesheets/_main.scss]: These padding props are not necessary, because this class is always used with `.comment-control-button`, which is overriding the padding.

File: client/app/reader/PdfListView.jsx [hunk 15]
```diff
@@ -33,74 +58,108 @@ export default class PdfListView extends React.Component {
...
+                selected={false}
+                page={comment.page}
+                onJumpToComment={this.props.onJumpToComment(comment)}
+                uuid={comment.uuid}
+                horizontalLayout={true}>
+                  {comment.comment}
+                </Comment>;
+            });
+
+            return <ul className="cf-no-styling-list" ariaLabel="Document comments">
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: This gives the following error: 

```
warning.js?8a56:36 Warning: Unknown prop `ariaLabel` on <ul> tag. Remove this prop from the element.
```

Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: It should be `aria-label`.

File: client/app/reader/PdfViewer.jsx [hunk 16]
```diff
@@ -276,5 +253,9 @@ PdfViewer.propTypes = {
   annotationStorage: PropTypes.object,
   doc: PropTypes.object,
   file: PropTypes.string.isRequired,
-  pdfWorker: PropTypes.string
+  label: PropTypes.string,
+  pdfWorker: PropTypes.string,
+  scrollToComment: PropTypes.number,
```
Comment (NickHeiner) [inline client/app/reader/PdfViewer.jsx]: When I load a viewer page with comments, I get a bunch of errors like:

```
Warning: Failed prop type: Invalid prop `scrollToComment` of type `object` supplied to `PdfViewer`, expected `number`.
    in PdfViewer (created by DecisionReviewer)
    in DecisionReviewer (created by Connect(DecisionReviewer))
    in Connect(DecisionReviewer) (created by Reader)
    in Provider (created by Reader)
    in Reader (created by BaseContainer)
    in div (created by BaseContainer)
    in BaseContainer
```

File: client/app/reader/PdfListView.jsx [hunk 17]
```diff
@@ -33,74 +58,108 @@ export default class PdfListView extends React.Component {
...
-                </a>
-              </span>
-            }
-          </span>;
-        }
+    return (row) => {
+      if (row && row.isComment) {
+        return [{
+          valueFunction: (doc) => {
+            let comments = this.props.annotationStorage.
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: I think we agreed to default to `const`:

https://dsva.slack.com/archives/C2ZAMLK88/p1492108961188361

**Comment by NickHeiner** [comment]
Looks like `key` isn't being set properly somewhere:

```
Each child in an array or iterator should have a unique "key" prop. Check the render method of `Comment`.
```

That comes up when I view a doc with comments.

File: client/app/components/Comment.jsx [hunk 18]
```diff
@@ -16,33 +17,76 @@ export default class Comment extends React.Component {
     this.props.onEditComment(this.props.uuid);
   }
 
+  controlButtons = () => {
+    let controlButtons = [];
+
+    if (this.props.onEditComment) {
+      controlButtons.push(<Button
+        name="edit"
+        classNames={["cf-no-padding-btn-link comment-control-button"]}
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: What is the benefit to doing this over:

```
className={"cf-no-padding-btn-link comment-control-button"}
```

Comment (NickHeiner) [inline client/app/components/Comment.jsx]: I also wouldn't mind factoring out this classname since it's shared with the delete button but nbd.

File: client/app/components/Comment.jsx [hunk 19]
```diff
@@ -16,33 +17,76 @@ export default class Comment extends React.Component {
     this.props.onEditComment(this.props.uuid);
   }
 
+  controlButtons = () => {
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: I like to put verbs in function names to emphasize that they do things. How about `getControlButtons`?

File: client/app/components/Comment.jsx [hunk 20]
```diff
@@ -51,15 +95,27 @@ export default class Comment extends React.Component {
...
         </div>
       </div>;
+    }
+
+    return commentToRender;
   }
 }
 
+Comment.defaultProps = {
+  onClick: () => {
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: I'm pretty sure you mean:

```js
onClick: _.noop
```

File: client/app/components/Pdf.jsx [hunk 21]
```diff
@@ -197,6 +198,7 @@ export default class Pdf extends React.Component {
         this.setState({
           numPages: pdfDocument.pdfInfo.numPages,
           pdfDocument,
+          setupPdfFile: file,
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: How about we start putting this into redux instead of adding more stuff to `this.state`?

File: client/app/components/Pdf.jsx [hunk 22]
```diff
@@ -212,6 +214,25 @@ export default class Pdf extends React.Component {
...
   }
 
+  onJumpToComment = (comment) => {
+    if (comment) {
+      let pageNumber = comment.page;
+      let yPosition = comment.y;
+
+      this.renderPage(pageNumber - 1).then(() => {
+        let pageElement = document.getElementById(`pageContainer${pageNumber}`);
+        let scrollWindow = document.getElementById('scrollWindow');
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: If we need to get the DOM node that corresponds to something we return from `render()`, I think it's better to use [refs](https://facebook.github.io/react/docs/refs-and-the-dom.html).

Staying away from IDs also helps us keep our component reusable, although in this particular case it might not make a huge difference.

File: client/app/components/Pdf.jsx [hunk 23]
```diff
@@ -212,6 +214,25 @@ export default class Pdf extends React.Component {
...
+        let height = (scrollWindow.getBoundingClientRect().bottom -
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: I wouldn't mind pulling this logic out into a unit test. We could have a standalone function that does this, which the component imports.

File: client/app/reader/DecisionReviewer.jsx [hunk 24]
```diff
@@ -31,6 +33,7 @@ export class DecisionReviewer extends React.Component {
       selectedLabels,
       sortBy: 'date',
       sortDirection: 'ascending',
+      scrollToComment: null,
```
Comment (NickHeiner) [inline client/app/reader/DecisionReviewer.jsx]: How about putting this in redux instead of local state?

File: client/app/reader/PdfListView.jsx [hunk 25]
```diff
@@ -1,11 +1,36 @@
...
 import DocumentListHeader from '../components/reader/DocumentListHeader';
+import _ from 'lodash';
+
+const NUMBER_OF_COLUMNS = 5;
 
 export default class PdfListView extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: How about using redux for this?

File: client/test/reader/DecisionReview-test.js [hunk 26]
```diff
@@ -229,7 +228,7 @@ describe('DecisionReviewer', () => {
         await pause();
 
         wrapper.find('#comment0').simulate('click');
-        expect(jumpTo.calledWith(commentId)).to.be.true;
+        expect(jumpTo.calledWith(sinon.match({ id: commentId }))).to.be.true;
```
Comment (NickHeiner) [inline client/test/reader/DecisionReview-test.js]: I always avoid chai assertions that end in a getter call (eg `to.do.something`) because if you misspell it, the result is a passing test:

```js
expect(false).to.be.treu
```

I've been burned by this enough times that I always do the method assertions. :smile:

File: client/test/reader/DecisionReview-test.js [hunk 27]
```diff
@@ -252,6 +251,28 @@ describe('DecisionReviewer', () => {
...
+
+      it('can jump to comment', asyncTest(async() => {
+        wrapper.find('#expand-2-comments-button').simulate('click');
+        wrapper.find('#button-jumpToComment').simulate('click');
+
+        let scrolledTo = sinon.spy(wrapper.getNode(), 'onCommentScrolledTo');
+
+        // verify the page is on the pdf view
+        expect(wrapper.text()).to.include('View all documents');
+        await pause();
```
Comment (NickHeiner) [inline client/test/reader/DecisionReview-test.js]: Why is this necessary?

File: client/app/components/Pdf.jsx (hunk 22 above)
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: That makes sense. But I'd rather not add that to this PR. Making a tech improvement ticket for it. #1608

File: client/app/reader/PdfListView.jsx (hunk 17 above)
Comment (mdbenjam) [inline client/app/reader/PdfListView.jsx]: Cool, I'll use that going forward.

File: client/test/reader/DecisionReview-test.js (hunk 27 above)
Comment (mdbenjam) [inline client/test/reader/DecisionReview-test.js]: Because the scrolling happens in the then of a promise, we have to have a pause.

File: client/app/components/Comment.jsx (hunk 18 above)
Comment (mdbenjam) [inline client/app/components/Comment.jsx]: There is no benefit, it's just the way that Button is currently built. It should definitely be changed.

File: client/app/components/Comment.jsx [hunk 28]
```diff
@@ -16,33 +17,72 @@ export default class Comment extends React.Component {
     this.props.onEditComment(this.props.uuid);
   }
 
+  getControlButtons = () => {
+    if (!this.props.onEditComment || !this.props.onDeleteComment) {
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: How about `if (!(this.props.onEditComment && this.props.onDeleteComment))`?

File: client/app/components/Comment.jsx [hunk 29]
```diff
@@ -16,33 +17,72 @@ export default class Comment extends React.Component {
...
+      return;
+    }
+
+    return <div>
+        <Button
+          name="edit"
+          classNames={["cf-btn-link comment-control-button"]}
+          onClick={this.onEditComment}>
```
Comment (NickHeiner) [inline client/app/components/Comment.jsx]: Did you mean `this.props`? (Same applies to `onDeleteComment` below.)

File: client/app/components/Pdf.jsx [hunk 30]
```diff
@@ -212,6 +217,23 @@ export default class Pdf extends React.Component {
     });
   }
 
+  onJumpToComment = (comment) => {
+    if (comment) {
+      let pageNumber = comment.page;
+      let yPosition = comment.y;
+
+      this.renderPage(pageNumber - 1).then(() => {
+        let height = (this.scrollWindow.getBoundingClientRect().bottom -
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: It would probably be more readable to factor out the `this.scrollWindow.getBoundingClientRect()` expression.

File: client/app/components/Pdf.jsx [hunk 31]
```diff
@@ -212,6 +217,23 @@ export default class Pdf extends React.Component {
...
+          this.scrollWindow.getBoundingClientRect().top);
+        let halfHeight = height / 2;
+
+        this.scrollWindow.scrollTop =
+          this.pageContainers[pageNumber - 1].getBoundingClientRect().top +
+          yPosition + this.scrollWindow.scrollTop - halfHeight;
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: I think it would be better to indent here since it's a continuation of the line before.

File: client/app/components/Pdf.jsx [hunk 32]
```diff
@@ -321,23 +329,38 @@ export default class Pdf extends React.Component {
...
         </div>);
     }
-
-    return <div id="scrollWindow" className="cf-pdf-scroll-view">
+    this.scrollWindow = null;
+
+    return <div
+      id="scrollWindow"
+      className="cf-pdf-scroll-view"
+      ref={(scrollWindow) => {
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Cool. Now that we have this, we can clean up the following in this component:

```js
        // Scroll to the correct location on the page
        document.getElementById('scrollWindow').scrollTop = scrollLocation;
```

File: client/app/components/Pdf.jsx [hunk 33]
```diff
@@ -347,6 +370,22 @@ export default class Pdf extends React.Component {
   }
 }
 
+const mapStateToProps = (state) => _.pick(state, 'currentRenderedFile');
+
+const mapDispatchToProps = (dispatch) => ({
+  handleSetCurrentRenderedFile(currentRenderedFile) {
+    dispatch({
+      type: Constants.SET_CURRENT_RENDERED_FILE,
+      payload: currentRenderedFile
```
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: I've been leaning towards a pattern of making `payload` always be an object that contains the actual payload information. That would make this:

```js
payload: { currentRenderedFile }
```

The idea is that we can trivially add new keys to the object, and existing code won't need to be changed if it's just reading the old keys. But it is marginally more complicated, and I don't feel strongly about it.

File: client/app/reader/index.jsx [hunk 34]
```diff
@@ -36,6 +36,16 @@ const readerReducer = (state = {}, action = {}) => {
         }
       }
     );
+  case Constants.SET_CURRENT_RENDERED_FILE:
+    return {
+      ...state,
+      currentRenderedFile: action.payload
```
Comment (NickHeiner) [inline client/app/reader/index.jsx]: My tendency has been to nest things more. For instance, in my PR I have something like this:

```js
{
  ui: {
    pdfList: {
      filters: {
        category: {
          procedural: true
        }
      },
      filterPositions: {
        category: {
          top: 210,
          right: 499.5625,
          bottom: 231,
          left: 479.78125,
          width: 19.78125,
          height: 21
        }
      },
      dropdowns: {
        category: true
      }
    }
  },
  documents: {
    '1': {
      id: 1,
      vbms_document_id: '1',
      category_procedural: true,
      category_medical: null,
      category_other: null,
      type: 'NOD',
      received_at: '2017-04-14',
      filename: 'filename.pdf',
      opened_by_current_user: true,
      receivedAt: '2017-04-14'
    }
}
```

I'm putting all documents under `documents`, and UI-related stuff that isn't persisted anywhere lives in `ui`.

File: client/app/components/Comment.jsx (hunk 28 above)
Comment (mdbenjam) [inline client/app/components/Comment.jsx]: De Morgan's law it up? Yeah, wasn't sure which one was clearer.

File: client/app/components/Comment.jsx (hunk 29 above)
Comment (mdbenjam) [inline client/app/components/Comment.jsx]: Nope, these are both methods in this class that wrap the props.

File: client/app/components/Pdf.jsx (hunk 31 above)
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: Hmm, I don't know if I agree with that. To me it looks weird if we indent another level. I feel like the one indent makes it clear.

File: client/test/reader/DecisionReview-test.js (hunk 26 above)
Comment (mdbenjam) [inline client/test/reader/DecisionReview-test.js]: I added an assert here. I'm going to keep the other expects for the time being (this PR is getting big), but will use asserts going forward.

File: client/app/components/Pdf.jsx [hunk 35]
```diff
@@ -70,7 +70,7 @@ export class Pdf extends React.Component {
       let pageNumber = index + 1;
       let canvas = document.getElementById(`canvas${pageNumber}`);
```
Comment (mdbenjam) [inline client/app/components/Pdf.jsx]: Let's take care of the other getElementById calls in another PR.

File: client/app/reader/DecisionReviewer.jsx [hunk 36]
```diff
@@ -375,9 +366,9 @@ const mapDispatchToProps = (dispatch) => ({
   onScrollToComment(scrollToComment) {
```
Comment (mdbenjam) [inline client/app/reader/DecisionReviewer.jsx]: When we do a bigger Redux refactor we should move this down to where it is needed.

File: client/app/reader/PdfListView.jsx [hunk 37]
```diff
@@ -205,11 +196,34 @@ export default class PdfListView extends React.Component {
   }
 }
 
+// Should be merged with documents when we finish integrating redux
+const mapStateToProps = (state) => {
+  return { reduxDocuments: _.get(state, 'documents') }
```
Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: We could also keep the names the same but namespace it under `redux`, like so:

```js
{
    redux: {
        documents: _.get(state, 'documents')
        // and add more stuff here over time before the migration is complete
    }
}
```

Comment (NickHeiner) [inline client/app/reader/PdfListView.jsx]: No need for `_.get` since `state` will always be defined.

File: client/app/reader/index.jsx [hunk 38]
```diff
@@ -46,6 +46,18 @@ const readerReducer = (state = {}, action = {}) => {
...
       scrollToComment: action.payload
     };
+  case Constants.TOGGLE_COMMENT_LIST:
+    return _.merge(
+      {},
+      state,
+      {
+        documents: {
+          [action.payload.docId]: {
+            listComments: action.payload.toggleState
```
Comment (NickHeiner) [inline client/app/reader/index.jsx]: So if we're always just toggling, then I would say don't include a payload, and just have the reducer invert the pre-existing value. That way, we can have bugs where the incorrect payload is sent.

File: client/test/reader/DecisionReview-test.js (hunk 27 above)
Comment (NickHeiner) [inline client/test/reader/DecisionReview-test.js]: How long does it pause for? Is it waiting for a condition, or are we just hoping that the scroll will be complete by the time the pause is done?

File: client/app/components/Pdf.jsx (hunk 31 above)
Comment (NickHeiner) [inline client/app/components/Pdf.jsx]: Ah, that's right, I see that it's already indented off of line 230. I missed that last time.

**Comment by NickHeiner** [comment]
I did a quick a11y scan of this and it seems good.

File: client/app/reader/index.jsx [hunk 39]
```diff
@@ -36,12 +45,48 @@ const readerReducer = (state = {}, action = {}) => {
         }
       }
     );
+  case Constants.SET_CURRENT_RENDERED_FILE:
+    return _.merge(
+      {},
+      state,
+      {
+        ui: {
+          pdf: {
```
Comment (NickHeiner) [inline client/app/reader/index.jsx]: This can also be:

```js
pdf: _.pick(action.payload, 'currentRenderedFile')
```

Same comment applies below.

**Comment by NickHeiner** [comment]
This looks good; only a few small comments left. Feel free to ping me on Slack to synchronously work through any of these issues.

Can we add a capybara test?

**Comment by NickHeiner** [comment]
I have spot checked that the unit tests fail when the code is broken.

**Comment by NickHeiner** [comment]
:+1: 36be03e looks good
